lib/config.c

```diff
@@ -428,14 +429,53 @@ walk_host_config(
                 log_internal(MTC_LOG_ERR, "%s: failed to get IPaddress\n", __func__);
                 return MTC_ERROR_CF_INVALID_FORMAT;
             }
-            c->common.host[c->common.hostnum].ip_address = inet_addr((char *) txt);
-            if (c->common.host[c->common.hostnum].ip_address == 0) 
+
+            const char *ip_address = (char *) txt;
+
+            struct addrinfo hints;
+            struct addrinfo *result;
+            memset(&hints, 0, sizeof hints);
+            hints.ai_family = AF_UNSPEC;
+            hints.ai_socktype = SOCK_DGRAM;
+            hints.ai_flags = 0;
+            hints.ai_protocol = 0;
+            hints.ai_canonname = NULL;
+            hints.ai_addr = NULL;
+            hints.ai_next = NULL;
+
+            errno = 0;
+            const int ret = getaddrinfo(ip_address, NULL, &hints, &result);
+            if (ret)
             {
-                log_internal(MTC_LOG_ERR, "%s: invalid IPaddress %s\n", __func__, txt);
+                log_internal(MTC_LOG_ERR, "%s: failed to get addr info of `%s` (info: %s). (sys %d)\n",
+                    __func__, ip_address, gai_strerror(ret), errno);
                 xmlFree(txt);
+                freeaddrinfo(result);
                 return MTC_ERROR_CF_INVALID_FORMAT;
             }
+
             xmlFree(txt);
+
+            struct sockaddr_storage *ss = &c->common.host[c->common.hostnum].sock_address;
+            switch (result->ai_family)
+            {
+                case AF_INET: {
+                    struct sockaddr_in *sa = (struct sockaddr_in *) ss;
+                    memcpy(&sa->sin_addr, &((struct sockaddr_in *) result->ai_addr)->sin_addr, sizeof sa->sin_addr);
+                    sa->sin_family = AF_INET;
+                } break;
+                case AF_INET6: {
+                    struct sockaddr_in6 *sa6 = (struct sockaddr_in6 *) ss;
+                    memcpy(&sa6->sin6_addr, &((struct sockaddr_in6 *) result->ai_addr)->sin6_addr, sizeof sa6->sin6_addr);
```

Comment:
To avoid all these casts should we have a union instead containing both sockaddr_in and sockaddr_in6, and then by selecting the member of the union you want you get the proper type?
It seems easy to typo one of these and accidentally cast to the wrong type.

(Obviously `bind` won't have the correct type, and would either need an additional member with sockaddr type, or cast to that type, both of which are not well defined behaviours in C: you can only access the member of the union using the same type you used to store, otherwise you need to memcpy to keep it fully general and supported across compilers).

Reply:
Perhaps it'd be better to add `-fno-strict-aliasing` to the compiler flags to avoid GCC's type-punning optimizer trying to do anything clever here and break the code. I'm not entirely sure you can write IPv4/IPv6 handling code without breaking GCC's type-punning rules (maybe with a few additional memcpy you can, but it is unavoidable that you store as `sockaddr_in6` or `sockaddr_storage` and read as `sockaddr`.)

Reply:
So, we replaced the `sockaddr_storage` variables using unions. :wink: I agree, it's better regarding maintenance and readability.

But I'm not sure it's mandatory to use the `-fno-strict-aliasing` flag, because the first attribute type of `sockaddr_in` and `sockaddr_in6` is the same (so it's a common initial part, and of course it's the start address), so we can cast to `sockaddr *` without risk (Ref: https://stackoverflow.com/questions/44245618/how-to-legally-use-type-punning-with-unions-to-cast-between-variations-of-struct#answer-44245922). Also we respect in this patch, the strict aliasing rules documented by gcc (https://gcc.gnu.org/onlinedocs/gcc-4.4.1/gcc/Optimize-Options.html#Type_002dpunning).

Reply:
As far as I know, sockaddr_storage is preferred for allocating storage, alignment, and accessing the `ss_family` member. There's some details here: https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/socket.h.html
This ensures that the code for allocation and determining the family doesn't change regardless of which address structures (in the future) are used.

So I think the preferred solution is to have a union that looks like this:

```
union {
    struct sockaddr sa;
    struct sockaddr_in sa_in;
    struct sockaddr_in6 sa_in6;
    struct sockaddr ss;
}
```

Determining the type happens via `foo.ss.ss_family`, accessing the individual types happens via `foo.sa_in` and `foo.sa_in6`, and calls to functions that take a struct sockaddr used `foo.sa`.

Reply:
We finished to change and fix the other comments with @benjamreis , I don't see a good advantage to add a new field like `sockaddr_storage ss` because the `ss_family` member is defined exactly using the same type on this struct and on `sockaddr`.  The unique impact of a new `sockaddr_storage` attribute is the replacement of  `ss_family` by `sa_family` and it doesn't change the behavior of our network calls, also `sockaddr_storage` is mainly useful when no union are used (and in this last case, there are many casts in a lot of functions).

Reply:
`sockaddr_storage` was what we used in our 1st implementation (before the union) and it resulted to many cast to `sockaddr` and so was not very readable.
After the we used the union we preferred to use `sockaddr` toi avoid the foremention casts since we do not use `sa_data` only `sa_family` so this is strictly identical as @Wescoeur said. :)